### ISSUE TITLE
margo pool size and noop ults

### ISSUE THREAD

**Issue by carns** [mochi-hpc/mochi-margo#6]
In GitLab by @jenkins on Mar 21, 2016, 15:10

In #5 @carns put in support for margo playing nicely in a "single thread" context (RPC runners in the same physical thread as margo progress). Here's the relevant code:

```c
ABT_pool_get_total_size(mid->progress_pool, &size);
if(size > 0)
{
    HG_Progress(mid->hg_class, mid->hg_context, 0);
    ABT_thread_yield();
}
else
{
    HG_Progress(mid->hg_class, mid->hg_context, 100);
}
```

Can't remember now, but there was at least some conversation about whether the `size > 0` check is the "right" one to make. The server example program would imply that it's not - the pool size is always `>= 1` due to the main thread being in the same pool as the one given to margo (causing the thread to revert to strictly polling). However, in my leveldb-hgbots example program, I create a new (empty) pool and thread for the program to run on, sleeping the main thread. Behavior there is as expected.

So how do we reconcile these two configurations? Does it suffice to add a parameter to `margo_init` indicating that there will be a ULT waiting on completion that it can safely ignore?

**Comment by carns** [comment]
In GitLab by @jenkins on Mar 21, 2016, 15:13

Oh I remember the conversation now - it was whether to use `ABT_pool_get_total_size` or `ABT_pool_get_size`. The latter doesn't count "blocked" ULTs, but isn't applicable here b/c we could have RPC ULTs doing things that might cause it to block (IOs, synchronizing with other ULTs/tasks).

**Comment by carns** [comment]
In GitLab by @carns on Mar 29, 2016, 10:46

The fundamental behavior that we want is for the caller (when Margo is executed in a single-threaded mode) to be able to sleep until some external entity (via an rpc, signal handler, thread, etc.) shuts down Margo.

From an API point of view we could do this by adding a margo_wait_for_finalize() call (not sure if that's the right name or not).  When the caller executes this it suspends the caller and tells the progress ULT that it is allowed to block on HG_Progress().  When someone else calls margo_finalize() it will return.

**Comment by carns** [comment]
In GitLab by @jenkins on Mar 29, 2016, 10:49

Perhaps a good bikeshed color is "margo_join"?

**Comment by carns** [comment]
In GitLab by @carns on Mar 29, 2016, 11:09

There is no margo_fork() though :)

**Comment by carns** [comment]
In GitLab by @jenkins on Mar 29, 2016, 11:11

Could always rename margo_init to margo_fork :)

**Comment by carns** [comment]
In GitLab by @carns on Mar 29, 2016, 20:43

Status changed to closed by commit b636e2535e3ad46bd685fc977e87cde1b323e2d0

**Comment by carns** [comment]
In GitLab by @carns on Mar 29, 2016, 20:45

I'll defer the fork/join discussion for another day; the original issue is resolved :)
